nts-pool-ke/src/lib.rs

```diff
@@ -209,10 +230,73 @@ async fn pool_key_exchange_server(
     }
 }
 
+async fn pick_nts_ke_server<'a>(
+    connector: &TlsConnector,
+    config::KeyExchangeServer { domain, port }: &'a config::KeyExchangeServer,
+    selected_algorithm: AeadAlgorithm,
+) -> Result<(&'a str, u16, ServerName), KeyExchangeError> {
+    info!("checking supported algorithms for '{domain}:{port}'");
+
+    let domain = domain.as_str();
+    let server_name = rustls::ServerName::try_from(domain)
+        .map_err(|_| std::io::Error::new(std::io::ErrorKind::InvalidInput, "invalid dnsname"))?;
+
+    let server_stream = match tokio::net::TcpStream::connect((domain, *port)).await {
+        Ok(server_stream) => server_stream,
+        Err(e) => return Err(e.into()),
+    };
+    let mut server_stream = connector
+        .connect(server_name.clone(), server_stream)
+        .await?;
+
+    info!("established connection to the server");
+
+    let supported_algorithms = supported_algorithms_request(&mut server_stream).await?;
+
+    info!("received supported algorithms from the NTS KE server");
+
+    if supported_algorithms
+        .iter()
+        .any(|(algorithm_id, _)| *algorithm_id == selected_algorithm as u16)
+    {
+        Ok((domain, *port, server_name))
+    } else {
+        Err(KeyExchangeError::NoValidAlgorithm)
+    }
+}
+
+async fn pick_nts_ke_servers<'a>(
+    connector: &TlsConnector,
+    servers: &'a [config::KeyExchangeServer],
+    selected_algorithm: AeadAlgorithm,
+    denied_servers: &[String],
+) -> Result<(&'a str, u16, ServerName), KeyExchangeError> {
+    for server in servers {
+        if denied_servers.contains(&server.domain) {
+            continue;
+        }
+
+        match pick_nts_ke_server(connector, server, selected_algorithm).await {
+            Ok(x) => return Ok(x),
+            Err(e) => match e {
```

Comment:
Shouldn't we also just `continue` if this server produces a `KeyExchangeError::NoValidAlgorithm` ? otherwise one ntske server that doesn't support our algorithm can frustrate the entire connection?